locksmith/src/websub/jobs/keys.ts

```diff
@@ -1,24 +1,76 @@
+import { Op } from 'sequelize'
 import { Key } from '../../graphql/datasource'
-import { Hook } from '../../models'
+import { Hook, ProcessedHookItem } from '../../models'
 import { TOPIC_KEYS } from '../topics'
 
-import { networkMapToFnResult, notifyHook } from '../helpers'
+import {
+  networkMapToFnResult,
+  notifyHook,
+  filterHooksByTopic,
+} from '../helpers'
 
-export async function notifyOfKeys(hooks: Hook[]) {
-  const subscribed = hooks.filter((hook) => {
-    const path = new URL(hook.topic).pathname
-    return TOPIC_KEYS.test(path)
+const FETCH_LIMIT = 25
+
+async function fetchUnprocessedKeys(network: number, page = 0) {
+  const keySource = new Key(network)
+  const keys = await keySource.getKeys({
+    first: FETCH_LIMIT,
+    skip: page ? page * FETCH_LIMIT : 0,
   })
 
-  const networkToLocksMap = await networkMapToFnResult((network) => {
-    const keySource = new Key(Number(network))
-    return keySource.getKeys({ first: 25 })
+  const keyIds = keys.map((key: any) => key.id)
+  const processedKeys = await ProcessedHookItem.findAll({
+    where: {
+      objectId: {
+        [Op.in]: keyIds,
+      },
+    },
   })
 
-  for (const hook of subscribed) {
-    const data = networkToLocksMap
-      .get(Number(hook.network))
-      .filter((key: any) => key.lock.id === hook.lock)
-    notifyHook(hook, data)
+  const unprocessedKeys = keys.filter(
+    (key: any) => !processedKeys.find((item) => item.objectId === key.id)
+  )
+  return unprocessedKeys
+}
+
+async function fetchAllUnprocessedKeys(network: number) {
+  let page = 0
+  const items = []
+
+  while (true) {
+    const keys = await fetchUnprocessedKeys(network, page)
+    // If empty, break the loop and return as there are no more new keys to process.
+    if (!keys.length) {
+      break
+    }
+    items.push(keys)
```

Comment:
Unfortunately that is still not the right approach and won't scale well.
For each batch of keys, we should notify the items in that batch, then save the batch and _finally_, fetch the next batch.
This way we won't miss any if for any reason something goes wrong somwhere.

Reply:
Got it.